R/rows-db.R

```diff
@@ -408,6 +468,160 @@ sql_rows_update_prep <- function(x, y, by) {
   )
 }
 
+#' @export
+#' @rdname rows-db
+sql_rows_patch <- function(x, y, by, ..., returning_cols = NULL) {
+  ellipsis::check_dots_used()
+  # FIXME: check here same src for x and y? if not -> error.
+  UseMethod("sql_rows_patch")
+}
+
+#' @export
+sql_rows_patch.tbl_SQLiteConnection <- function(x, y, by, ..., returning_cols = NULL) {
+  con <- dbplyr::remote_con(x)
+
+  p <- sql_rows_patch_prep(x, y, by)
+
+  sql <- paste0(
+    "WITH ", p$y_name, "(", p$y_columns_qq, ") AS (\n",
+    dbplyr::sql_render(y),
+    "\n)\n",
+    #
+    "UPDATE ", p$name, "\n",
+    "SET (", p$new_columns_qq, ") = (\n",
+    "SELECT ", p$new_columns_qual_qq, "\n",
+    "FROM ", p$y_name, "\n",
+    "WHERE (", p$compare_qual_qq, "))\n",
+    "WHERE EXISTS (SELECT * FROM ", p$y_name, " WHERE ", p$compare_qual_qq, ")",
+    sql_returning_cols(x, returning_cols)
+  )
+
+  glue::as_glue(sql)
+}
+
+#' @export
+`sql_rows_patch.tbl_Microsoft SQL Server` <- function(x, y, by, ..., returning_cols = NULL) {
+  con <- dbplyr::remote_con(x)
+
+  p <- sql_rows_patch_prep(x, y, by)
+
+  # https://stackoverflow.com/a/2334741/946850
+  sql <- paste0(
+    "WITH ", p$y_name, "(", p$y_columns_qq, ") AS (\n",
+    dbplyr::sql_render(y),
+    "\n)\n",
+    #
+    "UPDATE ", p$name, "\n",
+    "SET\n",
+    paste0(
+      "  ", unlist(p$new_columns_qq_list),
+      " = ", unlist(p$new_columns_qual_qq_list),
+      collapse = ",\n"
+    ),
+    "\n",
+    sql_output_cols(x, returning_cols),
+    "FROM ", p$name, "\n",
+    "  INNER JOIN ", p$y_name, "\n",
+    "  ON ", p$compare_qual_qq
+  )
+
+  glue::as_glue(sql)
+}
+
+#' @export
+sql_rows_patch.tbl_MariaDBConnection <- function(x, y, by, ..., returning_cols = NULL) {
+  con <- dbplyr::remote_con(x)
+
+  p <- sql_rows_patch_prep(x, y, by)
+
+  # https://stackoverflow.com/a/19346375/946850
+  sql <- paste0(
+    "UPDATE ", p$name, "\n",
+    "  INNER JOIN (\n", dbplyr::sql_render(y), "\n) AS ", p$y_name, "\n",
+    "  ON ", p$compare_qual_qq, "\n",
+    "SET\n",
+    paste0("  ", p$target_columns_qual_qq, " = ", p$new_columns_qual_qq, collapse = ",\n"),
+    sql_returning_cols(x, returning_cols)
+  )
+
+  glue::as_glue(sql)
+}
+
+#' @export
+sql_rows_patch.tbl_PqConnection <- function(x, y, by, ..., returning_cols = NULL) {
+  con <- dbplyr::remote_con(x)
+
+  p <- sql_rows_patch_prep(x, y, by)
+
+  # https://www.postgresql.org/docs/9.5/sql-update.html
+  sql <- paste0(
+    "WITH ", p$y_name, " AS (\n",
+    dbplyr::sql_render(y),
+    "\n)\n",
+    #
+    "UPDATE ", p$name, "\n",
+    "SET\n",
+    paste0(
+      "  ", unlist(p$new_columns_qq_list),
+      " = ", unlist(p$new_columns_qual_qq_list),
+      collapse = ",\n"
+    ),
+    "\n",
+    "FROM ", p$y_name, "\n",
+    "WHERE ", p$compare_qual_qq,
+    sql_returning_cols(x, returning_cols)
+  )
+
+  glue::as_glue(sql)
+}
+
+#' @export
+sql_rows_patch.tbl_duckdb_connection <- sql_rows_patch.tbl_SQLiteConnection
+
+sql_rows_patch_prep <- function(x, y, by) {
```

Comment:
Should we integrate this functionality into `sql_rows_update_prep()` or perhaps extract common parts?

Reply:
Probably, but I would do that in a separate PR and focus on implementing `rows_patch()` and `rows_upsert()` for now. What do you think?